lib/query/token-visitors.js

```diff
@@ -0,0 +1,188 @@
+const { isEmpty, compact, isNil, isFunction, isString, cloneDeep, castArray } = require("lodash");
+const Tokenizer = require("tokenize-this");
+const { operators } = require("../operators");
+
+class TokenListVisitor {
+    constructor(tokens) {
+        if (isString(tokens)) tokens = this.tokenize(tokens);
+        this.tokens = tokens;
+    }
+    tokenize(rawStr) {
+        const tokens = [];
+        const tokenizer = new Tokenizer();
+        tokenizer.tokenize(rawStr, (token) => {
+            if (tokens.length > 0) {
+                for (const combined of [
+                    tokens[tokens.length - 1] + token,
+                    tokens[tokens.length - 1] + ' ' + token
+                ]) {
+                    if (operators[combined]) {
+                        tokens[tokens.length - 1] = combined;
+                        return;
+                    }
+                }
+            }
+            tokens.push(`${token}`);
+        });
+        return tokens;
+    }
+    toString() {
+        return compact(this.tokens).join(' ');
+    }
+    findToken(value, isCaseSensitive = false, startIdx = 0) {
+        const lowerValue = value.toLowerCase();
+        for (let idx = startIdx; idx < this.tokens.length; idx++) {
+            const didMatch = isCaseSensitive
+                ? this.tokens[idx] === value
+                : this.tokens[idx].toLowerCase() === lowerValue;
+            if (didMatch) return new TokenItemVisitor(idx, this);
+        }
+        return null;
+    }
+    first() {
+        return new TokenItemVisitor(0, this);
+    }
+    last() {
+        return new TokenItemVisitor(this.tokens.length - 1, this);
+    }
+    _replaceRange(startIdx, endIdx, tokenList) {
+        const normalized = this._normalizeTokenList(tokenList);
+        this.tokens.splice(startIdx, endIdx - startIdx, ...normalized);
+        return new TokenRangeVisitor(startIdx, normalized.length, this);
+    }
+    _replaceAtIndex(i, tokenList) {
+        const normalized = this._normalizeTokenList(tokenList);
+        this.tokens.splice(i, 1, ...normalized);
+        return new TokenRangeVisitor(i, normalized.length, this);
+    }
+    _insertAtIndex(i, tokenList) {
+        const normalized = this._normalizeTokenList(tokenList);
+        if (normalized.length === 0) return null;
+        this.tokens.splice(i, 0, ...normalized);
+        return new TokenRangeVisitor(i, normalized.length, this);
+    }
+    _normalizeTokenList(tokenList) {
+        return compact(castArray(tokenList));
+    }
+    _selectRange(start, step, predicate) {
```

Comment:
can you please add comment explaining this logic?

Reply:
Yes, added now. Please lemme know if it is still unclear.